### ISSUE TITLE
JSON Extractor doesn't work with flattened JSON without any spaces inside

### ISSUE THREAD

**Issue by qarmin** [Graylog2/graylog2-server#7318]
Hi, 
I'm tried to extract JSON message sended by ESET with builtin JSON Extractor, but I'm unable to do this.

I'm getting from Eset ESMC syslogs log which have in message field this 
`{"event_type":"Audit_Event","ipv4":"www","hostname":"asfasfaas","source_uuid":"fase4b96","occured":"29-Jan-2020 09:50:07","severity":"Information","domain":"Domain group","action":"Logout","target":"wfe56615b","detail":"Logging out domain user 'asf.asf'.","user":"asf.asf","result":"Success"}`

This message should be properly extracted to multiple fields, but I have each time only `Nothing will be extracted` message. 

![12](https://user-images.githubusercontent.com/41945903/73353534-d33f7b00-4293-11ea-9434-a22bf720175c.png)

## Steps to Reproduce (for bugs)
1. Configure Eset 7 ESMC to send JSON Syslogs
2. Create input in Graylog
3. Try to create extractor in first message

## Your Environment

* Graylog Version: 3.2.0-3.beta.3
* Elasticsearch Version: 6.8.6-1
* MongoDB Version: 4.0.14-1.el7
* Operating System:  CentOS 7
* Browser version: Firefox 72.0.2

**Comment by kmerz** [comment]
I tried to reproduce your problem with the sample data you provided and it worked:

![Graylog - Edit extractor json](https://user-images.githubusercontent.com/448763/73434017-40134d80-4346-11ea-9bb4-e88df4d73116.png)

Could you please try it with the sample data (with a simple GELF or TCP input). If that works, can you please sample data more closely to your acutal data? Perhaps the problem is somewhere in the blacked out strings?

**Comment by qarmin** [comment]
How can I send manually messages to my server to test different messages and configuration?
I read manual, but I'm only found example with CURL and GELF which doesn't work for me
```
curl -XPOST http://graylog.example.org:12202/gelf -p0 -d '{"short_message":"Hello there", "host":"example.org", "facility":"test", "_foo":"bar"}'
```

**Comment by qarmin** [comment]
Can you try to check this JSON?
It is almost identically as original
```
{"event_type":"Audit_Event","ipv4":"182.12.1.242","hostname":"romna.sqe.feg","source_uuid":"5ab3459d-d6e5-480b-8a1e-8811d62e4b96","occured":"30-Jan-2020 12:42:53","severity":"Information","domain":"Domain group","action":"Login attempt","target":"763d4a3f-98fe-4697-93a4-5fd36e73615b","detail":"Authenticating domain user 'roman.koudelka'.","user":"","result":"Success"}
```

**Comment by kmerz** [comment]
This is how I send the message to my gelf http input listening on port 12201:

`curl -XPOST http://localhost:12201/gelf -p0 -d '{"short_message":"{\"event_type\":\"Audit_Event\",\"ipv4\":\"182.12.1.242\",\"hostname\":\"romna.sqe.feg\",\"source_uuid\":\"5ab3459d-d6e5-480b-8a1e-8811d62e4b96\",\"occured\":\"30-Jan-2020 12:42:53\",\"severity\":\"Information\",\"domain\":\"Domain group\",\"action\":\"Login attempt\",\"target\":\"763d4a3f-98fe-4697-93a4-5fd36e73615b\",\"detail\":\"Authenticating domain user 'roman.koudelka'.\",\"user\":\"\",\"result\":\"Success\"}", "host":"hulud.net", "facility":"test", "_foo":"bar"}'`

And here the JSON Extractor result:
![Graylog - Edit extractor json (1)](https://user-images.githubusercontent.com/448763/73452117-7c0bda00-4369-11ea-9919-ffdfb0c9fe77.png)

Do you see anything in your graylog server logs?

**Comment by qarmin** [comment]
> gelf http input
I was using gelf http tcp, and that is why test messages not working.

So, Gelf HTTP works perfectly,  I don't need even enable Flatten Structures, because everything extract correctly.
But with Syslog UDP, I still can't use JSON extractor.

I tested it on two exactly same messages, so seems that something is wrong with Syslog UDP

**Comment by qarmin** [comment]
This also doesn't work with Syslog TCP

**Comment by kmerz** [comment]
Hello @qarmin,

the input does not matter. The content of the field is only important:
![Graylog - New extractor for input localhost](https://user-images.githubusercontent.com/448763/73534699-07e13d00-4422-11ea-9f2c-26f86b49b500.png)

This is the result from a Syslog UDP Input.

I had problems in the beginning since the json was not escapted correctly. You can try it yourself with this: 
```
logger '{"event_type":"Audit_Event","ipv4":"182.12.1.242","hostname":"romna.sqe.feg","source_uuid":"5ab3459d-d6e5-480b-8a1e-8811d62e4b96","occured":"30-Jan-2020 12:42:53","severity":"Information","domain":"Domain group","action":"Login attempt","target":"763d4a3f-98fe-4697-93a4-5fd36e73615b","detail":"Authenticating domain user 'roman.koudelka'.","user":"","result":"Success"}'
```

**Comment by kmerz** [comment]
@qarmin how do you extract the json part from the message?

**Comment by qarmin** [comment]
I already have message showed as JSON by default(I don't change anything)

Message from top is from ESET and from below it comes from Logger centos command
![DDD](https://user-images.githubusercontent.com/41945903/73540092-44676580-442f-11ea-9f03-e500bffd9137.png)

**Comment by kmerz** [comment]
Can you take a look in your graylog server.log file and look for extractor errors when click on `Try`?
Also can you check if any other extractor is configured on that input? Sometimes the output of one extractor can make it impossible for the JSON parser to read the string.

**Comment by qarmin** [comment]
I created new a new input only for Eset and restated server but still JSON extractor not working and nothing is printed to /var/log/graylog-server/server.log when I click at Try button

**Comment by qarmin** [comment]
I found a workaround, and probably cause of this problem
As workaround I enabled Full Messages and created Regex extractor to extract all JSON in format `{json:json}` - I used this regex for search `[^\{]+\{` and this for replace `\{`
![ESET](https://user-images.githubusercontent.com/41945903/73639774-4fa7d480-466d-11ea-95ab-da4a0eba8835.png)
And JSON extractor for this works perfectly

I copied to WineMerge text from 'message' and 'jsonFix'(created by me in previous step) an I found that this messages are different
![ESET2](https://user-images.githubusercontent.com/41945903/73639887-8aaa0800-466d-11ea-8cb2-614789ee3a11.png)

I copied text `{"event_type"` to site https://www.browserling.com/tools/text-to-ascii and I got different results
- Default message - `239 187 191 123 34 101 118 101 110 116 95 116 121 112 101 34`
- JsonFix - `123 34 101 118 101 110 116 95 116 121 112 101 34`

So at the begging there are `239 187 191` which seems to broke JSON Extractor.

**Comment by kmerz** [comment]
@qarmin that is a really helpful information! It makes sense that JSON can't parse this strange character. I had a glimpse on ESET and saw that some of their applications support different SYSLOG fromats. Could you try another one? And to document the behavior here, what is currently selected?

**Comment by qarmin** [comment]
There are several options in Eset, but seems that nothing works

- [x] - means that I tested it

Format
- [x] Syslog
- [ ] BSD - Graylog doesn't seems to support this

Transport
- [x] TCP
- [x] UDP
- [ ] TLS

Octet-counted framing
- [x] ON
- [x] OFF

Exported logs format:
- [x] JSON
- [x] LEEF - Logs are very strange and hard to parse, and still contains this strange numbers
![2](https://user-images.githubusercontent.com/41945903/73658051-dc658900-4693-11ea-9ff7-4afc0c738cdb.png)

Eset settings page

![ESS](https://user-images.githubusercontent.com/41945903/73658109-061eb000-4694-11ea-9e75-f9e0c3062153.png)

**Comment by kmerz** [comment]
As far as I know we so support BSD formatted syslog messages. Just for the sake of completion of this Issue could you try BSD as well?

Otherwise you are right with everything you wrote graylog has problems converting json if any content precedes the first `{` character.

Another workaround would be to create your extractors as pipeline rules.

But the actual bug is hard to fix without a way to have the ESET server running for our self. So I will leave this bug open as a reminder.

**Comment by qarmin** [comment]
Eset doesn't add this strange characters to BSD logs, but still I can't use JSON Extractor default message, because I have before first { - "eset..... ERAServer[1144]: "

![QQWE](https://user-images.githubusercontent.com/41945903/73726065-fe601980-472e-11ea-946c-2ae770900204.png)

**Comment by qarmin** [comment]
This numbers seems to be BOM of utf-8 - https://en.wikipedia.org/wiki/Byte_order_mark#UTF-8
I fill bug report to Eset, but I get info that formatting of syslog depends on the Syslog server.

**Comment by maniel** [comment]
i noticed it sometime ago, i asked on Graylog Comunity about it [here](https://community.graylog.org/t/json-extractor-not-working/6648), it seems like ESET sends BOM at the beginning of the json message and it breaks Graylog extractor, the workaround i found out is included in the linked thread
